### PR TITLE
feat(databind): improve validation error messages with location and context

### LINKED ISSUE #205: Improve exception reporting for null values on optional field values
### User Story

As a developer using this library and related tooling, when encountering an optional field value, I would like a more precise error message and controlled error behavior when such a null value is visited. I do not want a null pointer exception.

### Goals

- [ ] Improve error handling for users of metaschema-java based tools
- [ ] Gracefully fail in the occurrence of null values for a fireld

### Dependencies

N/A

### Acceptance Criteria

- [ ] All website and readme documentation affected by the changes in this issue have been updated.
- [ ] A Pull Request (PR) is submitted that fully addresses the goals of this User Story. This issue is referenced in the PR.
  - [ ] Remove the `ObjectUtils.requireNonNull()` check on https://github.com/metaschema-framework/metaschema-java/blob/ebdf350362e3c75f2fb39969a407089464a9b7c0/databind/src/main/java/gov/nist/secauto/metaschema/databind/model/IBoundDefinitionModelFieldComplex.java#L77 and other applicable areas in the code base.
  - [ ] Properly implement null-checking and throw the correct exception type in place of the NPE generated from the replaced null check above
  - [ ] Implement tests to check for correctness of implementation and gaps as approprriate
- [ ] The CI-CD build process runs without any reported errors on the PR. This can be confirmed by reviewing that all checks have passed in the PR.

### Revisions

_No response_

### LINKED ISSUE #595: Use format-appropriate names in validation error messages
## Summary

Validation error messages currently use JSON property names (`getJsonName()`) for identifying missing or invalid fields, even when parsing XML content. This can be confusing for users working with XML-only scenarios.

## Current Behavior

When parsing XML and a required field is missing, the error message references the JSON property name:
```
Missing required property: some-json-name
```

## Proposed Behavior

Error messages should adapt based on the parsing context:
- **JSON/YAML parsing**: Use JSON property name
- **XML parsing**: Use XML element/attribute name

Example for XML:
```
Missing required element: some-xml-element
```

## Affected Code

- `AbstractProblemHandler.java` - `isRequiredAndMissingDefault()` and related methods use `getJsonName()`
- `DefaultXmlProblemHandler.java` - Should use XML names
- `DefaultJsonProblemHandler.java` - Should continue using JSON names

## Implementation Notes

- The `IBoundProperty` interface provides both `getJsonName()` and `getXmlQName()` methods
- Format context would need to be passed through the problem handler hierarchy
- Consider whether to use qualified names or local names for XML

## Related

This was identified during PR #593 code review as a nice-to-have improvement.

### LINKED ISSUE #596: Improve validation error messages with location and contextual information
## Summary

Validation error messages during deserialization should provide sufficient detail to support user debugging of content. Currently, some error messages lack location information (line/column) and contextual details that would help users pinpoint and understand problems.

### Required Field Validation (needs improvement)

The new required field validation (`DESERIALIZE_VALIDATE_REQUIRED_FIELDS`) produces messages like:

```
Missing required property in root: required-flag
Missing required properties in root: required-flag, required-field, required-assembly
```

**Issues:**
- No file location (line/column numbers)
- No XPath or element path context
- Generic "property" term doesn't distinguish flag/field/assembly
- Uses JSON names only (may confuse XML users)

### Other Parsing Errors (good examples to follow)

XML malformed data errors include location:
```
Malformed data 'invalid-number' in 'file:///data.xml' at 3:10. For input string: "invalid-number"
```

JSON missing property errors include location:
```
Failed to find property with name 'required-field' in 'file:///data.json' at '2:5'.
```

### Format Inconsistency

| Error Source | Location Format |
|--------------|-----------------|
| XML parsing | `at [line]:[column]` |
| JSON parsing | `at '[line]:[column]'` (with quotes) |
| Required field validation | None |

### Option A: Minimal Enhancement (add location only)

Add file and line/column to existing message format:

```
Missing required property in root: required-flag
  at file:///example.xml line 3
```

**Pros:**
- Simple implementation
- Maintains current message style
- Low risk of breaking existing error handling

**Cons:**
- Still uses generic "property" terminology
- No element path context for deeply nested structures

### Option B: Detailed Context (location + type + path)

Provide richer context with element types and paths:

```
Missing required flag 'required-flag' in assembly 'root'
  Location: file:///example.xml at 3:45
  XPath: /root
```

**Pros:**
- Clear distinction between flags, fields, and assemblies
- Path helps locate issues in complex documents
- More actionable for users

**Cons:**
- More complex implementation
- Requires tracking parser position during validation

### Option C: Multi-line Rich Format (maximum context)

Maximum debugging information:

```
Validation Error: Missing required properties
  File: example.xml
  Location: line 3, column 45
  Parent: assembly 'root' (namespace: http://example.com/ns)
  Missing:
    - flag 'required-flag' (XML: @required-flag)
    - field 'required-field' (XML: <required-field>)
  XPath: /root
```

**Pros:**
- Maximum debuggability
- Shows both JSON and XML representations
- Groups multiple errors together

**Cons:**
- Verbose output
- Significant refactoring required
- May be overkill for simple cases

## Recommendation

**Option B (Detailed Context)** provides the best balance:
- Includes location information (file, line, column)
- Distinguishes between property types (flag/field/assembly)
- Provides path context for navigation
- Reasonable implementation complexity

## Implementation Considerations

1. **Location Tracking**: The problem handler receives instances after parsing completes, so location information must be captured during parsing and associated with validation errors.

2. **Consistent Format**: Standardize location format across XML and JSON (e.g., always use `at line:column` without quotes).

3. **Choice Group Context**: When reporting missing required fields in choice groups, indicate which alternatives were expected.

4. **Internationalization**: Consider whether error messages should be externalizable for i18n.

## Acceptance Criteria

- [ ] Required field validation errors include file URI
- [ ] Required field validation errors include line and column numbers
- [ ] Error messages distinguish between flags, fields, and assemblies
- [ ] Error messages include element/property path for context
- [ ] Location format is consistent between XML and JSON parsing
- [ ] Choice group validation errors indicate available alternatives
- [ ] Existing parsing error formats are reviewed for consistency

## Related

- PR #593 - Added required field validation with current error format
- Issue #594 - Choice instance support for annotation-based bindings

### PR DESCRIPTION
**Pull request by david-waltermire**
## Summary

This PR improves validation error messages during deserialization to provide better debugging context for users.

### Changes

- **Format-appropriate names**: Use XML element/attribute names when parsing XML, JSON property names when parsing JSON
- **Location information**: Include file URI, line number, and column number in validation errors
- **Path context**: Include document path (XPath-style for XML, JSON Pointer-style for JSON)
- **Property type identification**: Distinguish between flags, fields, and assemblies in error messages
- **Null field value handling**: Replace NPE with informative validation errors

### Example Error Messages

Before:
```
Missing required property in root: required-flag
```

After:
```
Missing required flag 'control-id' in assembly 'catalog'
  Location: file:///example.xml at 15:8
  Path: /catalog/metadata
```

### Implementation

- New `ValidationContext` class bundles parsing context (location, path, format)
- New `PathTracker` utility for lightweight path tracking during parsing
- Updated `AbstractProblemHandler` with enhanced error formatting
- Updated readers (`MetaschemaJsonReader`, `MetaschemaXmlReader`) to capture and pass context

## Test Plan

- [ ] Tests for format-appropriate names (XML vs JSON)
- [ ] Tests for location information in error messages
- [ ] Tests for path context
- [ ] Tests for property type distinction (flag/field/assembly)
- [ ] Tests for null field value handling
- [ ] Tests for edge cases (root level, no URI, choice groups, defaults)
- [ ] Full build verification: `mvn clean install -PCI -Prelease`

## Related Issues

Closes #595
Closes #596
Closes #205

## Summary by CodeRabbit

* **New Features**
  * Richer, context-aware validation errors: include source/file name, line/column when available, parsing path, and format-appropriate wording (XML vs JSON/YAML); safer handling of missing or null fields to avoid parse crashes.

* **Documentation**
  * Added comprehensive PRD for validation errors, user-facing error terminology guidance, updated PRD/workflow conventions, PR feedback workflow, and git submodule initialization instructions.

* **Tests**
  * New test suite validating format-specific messages, location/path context, edge cases, and null-field scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/PathTracker.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  /**
+   * Pop the most recent segment from the path.
+   *
+   * @return the removed segment, or null if the path was empty
+   */
+  public String pop() {
+    return segments.poll();
+  }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/io/PathTracker.java:60]: Addressed in commit 6b71fb4 by adding @Nullable annotation to pop() return type.

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/xml/DefaultXmlDeserializer.java [hunk 2]
```diff
@@ -150,10 +155,13 @@ protected final IDocumentNodeItem deserializeToNodeItemInternal(Reader reader, U
 
   @Override
   public final CLASS deserializeToValueInternal(Reader reader, URI resource) throws IOException {
+    // Handle null URI gracefully - use a synthetic URI for parsing
+    URI effectiveResource = resource != null ? resource : UNKNOWN_SOURCE;
+
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/io/xml/DefaultXmlDeserializer.java]: Addressed in commit e45cabea. Aligned with the @NonNull contract from AbstractDeserializer - removed null-handling code entirely instead of adding @Nullable.

File: databind/src/test/java/gov/nist/secauto/metaschema/databind/io/ValidationErrorMessageTest.java [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+      IOException exception = assertThrows(IOException.class, () -> {
+        deserializer.deserialize(new StringReader(xml), null);
+      });
+
+      String message = exception.getMessage();
+      assertNotNull(message, "Exception message should not be null");
+      // Should still have useful error message even without URI
+      assertTrue(message.toLowerCase().contains("required") || message.toLowerCase().contains("missing"),
+          "Error message should indicate missing required property: " + message);
+    }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/test/java/gov/nist/secauto/metaschema/databind/io/ValidationErrorMessageTest.java:383]: Addressed in commit e45cabea. Aligned both deserializers to the @NonNull contract from AbstractDeserializer - removed null-handling code and updated the test to use a valid URI instead of null.

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/xml/DefaultXmlDeserializer.java [hunk 4]
```diff
@@ -149,11 +155,14 @@ protected final IDocumentNodeItem deserializeToNodeItemInternal(Reader reader, U
...
+  public final CLASS deserializeToValueInternal(Reader reader, @Nullable URI resource) throws IOException {
+    // Handle null URI gracefully - use a synthetic URI for parsing
+    URI effectiveResource = resource != null ? resource : UNKNOWN_SOURCE;
+
     // doesn't auto close the underlying reader
     try (AutoCloser<XMLEventReader2, XMLStreamException> closer = AutoCloser.autoClose(
-        newXMLEventReader2(resource, reader), XMLEventReader::close)) {
-      return parseXmlInternal(closer.getResource(), resource);
+        newXMLEventReader2(effectiveResource, reader), XMLEventReader::close)) {
+      return parseXmlInternal(closer.getResource(), effectiveResource);
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/io/xml/DefaultXmlDeserializer.java]: Addressed in commit e45cabea. Aligned with the @NonNull contract from AbstractDeserializer - removed null-handling code and associated Javadoc.

File: databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/DefaultJsonDeserializer.java [hunk 5]
```diff
@@ -129,9 +134,12 @@ private MetaschemaJsonReader newMetaschemaJsonReader(
   @Override
   protected INodeItem deserializeToNodeItemInternal(@NonNull Reader reader, @NonNull URI documentUri)
       throws IOException {
+    // Handle null URI gracefully - use a synthetic URI for parsing
+    URI effectiveUri = documentUri != null ? documentUri : UNKNOWN_SOURCE;
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply databind/src/main/java/gov/nist/secauto/metaschema/databind/io/json/DefaultJsonDeserializer.java]: Addressed in commit e45cabea. Aligned with the @NonNull contract from AbstractDeserializer by removing the null-check fallback code. The @NonNull annotations are now correct and match the actual behavior.

File: .claude/commands/address-pr-feedback.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+Retrieve all review comments on the PR:
+
+```bash
+# Get PR number
+PR_NUMBER=$(gh pr list --head $(git branch --show-current) --json number --jq '.[0].number')
+
+# Fetch all review comments with context
+gh api repos/{owner}/{repo}/pulls/$PR_NUMBER/comments \
+  --jq '.[] | {id: .id, path: .path, line: .line, body: .body[0:300], user: .user.login, in_reply_to_id: .in_reply_to_id}'
+```
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply .claude/commands/address-pr-feedback.md]: @coderabbitai Addressed in commit 279e60224. Replaced {owner}/{repo} placeholders with dynamic REPO variable extracted via `gh repo view --json nameWithOwner`.

File: .claude/commands/address-pr-feedback.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+echo "=== General PR Comments ==="
+gh api repos/$REPO/issues/$PR_NUMBER/comments \
+  --jq '.[] | {id: .id, type: "issue", body: .body[0:300], user: .user.login}'
+```
+
+**Comment Types:**
+| Type | API Endpoint | Description |
+|------|--------------|-------------|
+| Review comments | `/pulls/{id}/comments` | Line-specific comments on files |
+| Issue comments | `/issues/{id}/comments` | General PR-level discussion |
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply .claude/commands/address-pr-feedback.md:48]: @coderabbitai Addressed in commit 810f5c768. Added blank line before the table.

File: .claude/commands/address-pr-feedback.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+   - Implement the fix
+   - Verify tests pass
+5. **Verify the fix** compiles and all tests pass
+
+Track progress with a checklist:
+- [ ] Comment 1: [description] - [status]
+- [ ] Comment 2: [description] - [status]
+- ...
+
+**Batch related fixes** into a single commit when possible.
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply .claude/commands/address-pr-feedback.md:96]: @coderabbitai This is grammatically correct as-is. "Batch" is the imperative verb (meaning "group together"), and "related" is an adjective modifying "fixes". It reads "Batch [the] related fixes into a single commit" - similar to "Group related items together". No hyphenation needed.

File: .claude/commands/address-pr-feedback.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+
+# Get the reviewer's login from the comment
+REVIEWER_LOGIN=$(gh api repos/$REPO/pulls/$PR_NUMBER/comments/<comment_id> --jq '.user.login')
+
+# Reply with appropriate mention
+gh api repos/$REPO/pulls/$PR_NUMBER/comments \
+  -X POST \
+  -f body="@$REVIEWER_LOGIN Addressed in commit abc1234. <explanation>" \
+  -f in_reply_to=<original_comment_id>
+```
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply .claude/commands/address-pr-feedback.md:133]: @coderabbitai Addressed in commit e6c10310b. Updated to use the dedicated /comments/{id}/replies endpoint. Note: the previous in_reply_to approach also works per GitHub API docs, but this is cleaner.
